src/twisted/protocols/policies.py

```diff
@@ -365,15 +365,15 @@ def startFactory(self):
         self.peerConnections = {}
 
     def buildProtocol(self, addr):
-        peerHost = addr[0]
+        peerHost = addr.host
```

Comment:
Notice that `buildProtocol` (https://docs.twisted.org/en/stable/api/twisted.internet.interfaces.IProtocolFactory.html#buildProtocol)'s `addr` argument is only an `IAddress` (https://docs.twisted.org/en/stable/api/twisted.internet.interfaces.IAddress.html) and `IAddress` has no attributes (nor is it indexable - so the code was equally broken before).

Without narrowing the type at runtime, there's not much you can do with `addr` - notice how many implementations of `IAddress` won't have a "host".

I don't know what the best way to make this work is.  One way that might work, though, would be to introduce an `IHasPeerIdentity` interface with a `getPeerIdentity` method and then write a lot of adapters from the different `IAddress` implementation types to that new interface.  Then each address type can be dealt with separately while this factory remains general, and the system is still open to new address types (someone will just need to write an adapter if they want a new address type to work).

This still leaves you open to a runtime error if you try to use this factory with an address type that can't be adapted to the new interface, though.  That could be turned into a different kind of error by supplying a default implementation of adaption fails - one that either treats all addresses of an unknown type as having the same identity (fail non-functional ("fail open")) or as each having a different identity (fail functional ("fail closed")).

It doesn't *seem* like the way factory types and address types are currently laid out allows for anything other than runtime detection of this case... but maybe someone who remains more actively engaged with development will think of a better idea.

Reply:
To avoid the repeated runtime checks, we could do a one-time check with the class, looking at type annotations and warning if they're not present.